listen.go

```diff
@@ -149,78 +138,112 @@ func (l *listener) handleIncoming() {
 	defer wg.Done()
 
 	for {
-		maconn, err := l.Listener.Accept()
+		conn, err := l.Listener.Accept()
 		if err != nil {
 			if l.catcher.IsTemporary(err) {
 				continue
 			}
-
-			l.incoming <- connErr{err: err}
+			l.incoming <- connOrErr{err: err}
 			return
 		}
 
-		log.Debugf("listener %s got connection: %s <---> %s", l, maconn.LocalMultiaddr(), maconn.RemoteMultiaddr())
-
-		if l.filters != nil && l.filters.AddrBlocked(maconn.RemoteMultiaddr()) {
-			log.Debugf("blocked connection from %s", maconn.RemoteMultiaddr())
-			maconn.Close()
+		if l.filters != nil && l.filters.AddrBlocked(conn.RemoteMultiaddr()) {
+			log.Debugf("blocked connection from %s", conn.RemoteMultiaddr())
+			conn.Close()
 			continue
 		}
-		// If we have a wrapper func, wrap this conn
-		if l.wrapper != nil {
-			maconn = l.wrapper(maconn)
-		}
+
+		log.Debugf("listener %s got connection: %s <---> %s", l, conn.LocalMultiaddr(), conn.RemoteMultiaddr())
 
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			if l.protec != nil {
-				pc, err := l.protec.Protect(maconn)
+
+			ctx, cancel := context.WithTimeout(l.ctx, ConnAcceptTimeout)
+			defer cancel()
+
+			done := make(chan struct{})
+			go func() {
+				defer close(done)
+
+				if l.protec != nil {
+					pc, err := l.protec.Protect(conn)
+					if err != nil {
+						conn.Close()
+						log.Warning("protector failed: ", err)
+						return
+					}
+					conn = pc
+				}
+
+				// If we have a wrapper func, wrap this conn
+				if l.wrapper != nil {
+					conn = l.wrapper(conn)
+				}
+
+				var stream io.ReadWriteCloser
+				switch conn := conn.(type) {
+				case tpt.DuplexConn:
+					stream = conn
+				case tpt.MultiplexConn:
+					stream, err = conn.AcceptStream()
+					if err != nil {
+						conn.Close()
+						log.Warning("accepting stream failed: ", err)
+						return
+					}
+					defer stream.Close()
+				}
+
+				if _, _, err := l.mux.Negotiate(stream); err != nil {
```

Comment:
Definitely unrelated to this PR, but could we check that the protocol chosen here is secio, and error out otherwise? It would be helpful if nodes on older code don't behave weirdly when we start supporting tls1.3